internal/k8s/statusaddress.go

```diff
@@ -14,86 +14,96 @@
-	Status       v1.LoadBalancerStatus
-	IngressClass string
+type StatusAddressUpdater struct {
+	Logger        logrus.FieldLogger
+	LBStatus      v1.LoadBalancerStatus
+	IngressClass  string
+	StatusUpdater StatusUpdater
 }
 
-func (s *IngressStatusUpdater) OnAdd(obj interface{}) {
+func (s *StatusAddressUpdater) OnAdd(obj interface{}) {
 
 	ing := obj.(*v1beta1.Ingress).DeepCopy()
+	ing.SetGroupVersionKind(v1beta1.SchemeGroupVersion.WithKind("ingress"))
 	if !annotation.MatchesIngressClass(ing, s.IngressClass) {
 		s.Logger.
 			WithField("name", ing.GetName()).
 			WithField("namespace", ing.GetNamespace()).
 			WithField("ingress-class", annotation.IngressClass(ing)).
+			WithField("configured-ingress-class", s.IngressClass).
 			Debug("unmatched ingress class, skip status update")
 		return
 	}
 
-	ing.Status.LoadBalancer = s.Status
-	_, err := s.Client.NetworkingV1beta1().Ingresses(ing.GetNamespace()).UpdateStatus(context.TODO(), ing, metav1.UpdateOptions{})
-	if err != nil {
-		s.Logger.
-			WithField("name", ing.GetName()).
-			WithField("namespace", ing.GetNamespace()).
-			WithError(err).Error("unable to update status")
+	m := mutator{
+		lbstatus: s.LBStatus,
 	}
+
+	s.StatusUpdater.SetStatus(
+		ing.GetName(),
+		ing.GetNamespace(),
+		v1beta1.SchemeGroupVersion.WithResource("ingresses"),
+		m,
+	)
 }
 
-func (s *IngressStatusUpdater) OnUpdate(oldObj, newObj interface{}) {
+func (s *StatusAddressUpdater) OnUpdate(oldObj, newObj interface{}) {
 
-	oldIng := oldObj.(*v1beta1.Ingress).DeepCopy()
 	newIng := newObj.(*v1beta1.Ingress).DeepCopy()
-
-	// We need to only act when things come *into* our ingressclass scope. When they fall out, we don't care about them any
-	// more, and it's the new controller's job to fix things.
-	// Note that this also handles the case where someone deletes the annotation
-	if !annotation.MatchesIngressClass(oldIng, s.IngressClass) && annotation.MatchesIngressClass(newIng, s.IngressClass) {
+	newIng.SetGroupVersionKind(v1beta1.SchemeGroupVersion.WithKind("ingress"))
+	if annotation.MatchesIngressClass(newIng, s.IngressClass) {
 		// Add status because we started matching ingress-class.
 		s.Logger.
 			WithField("name", newIng.GetName()).
 			WithField("namespace", newIng.GetNamespace()).
 			WithField("ingress-class", annotation.IngressClass(newIng)).
+			WithField("configured-ingress-class", s.IngressClass).
 			Debug("Updated Ingress is in scope, updating")
-		newIng.Status.LoadBalancer = s.Status
-		_, err := s.Client.NetworkingV1beta1().Ingresses(newIng.GetNamespace()).UpdateStatus(context.TODO(), newIng, metav1.UpdateOptions{})
-		if err != nil {
-			s.Logger.
-				WithField("name", newIng.GetName()).
-				WithField("namespace", newIng.GetNamespace()).
-				WithError(err).Error("unable to update status")
+		m := mutator{
+			lbstatus: s.LBStatus,
 		}
+
+		s.StatusUpdater.SetStatus(
+			newIng.GetName(),
+			newIng.GetNamespace(),
+			v1beta1.SchemeGroupVersion.WithResource("ingresses"),
+			m,
```

Comment:
Rather than a local struct type, you can adapt a function in place:

```Go
k8s.StatusMutatorFunc(func (obj interface{}) interface {
  ing := obj.(*v1beta1.Ingress).DeepCopy()
  ing.Status.LoadBalancer = s.LBStatus
  return ing
})
```

Reply:
Agreed, but I'll still need a local function so that there's only one place where that mutation function is kept.